pmtiles.cpp

```diff
@@ -0,0 +1,418 @@
+#include <stdio.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <zlib.h>
+#include "pmtiles.hpp"
+#include <protozero/varint.hpp>
+#include <iostream>
+#include <sstream>
+#include "mvt.hpp"
+#include "write_json.hpp"
+#include "version.hpp"
+
+pmtilesv3 *pmtilesv3_open(const char *filename, char **argv, int force, const char* tmpdir) {
+	pmtilesv3 *outfile = new pmtilesv3;
+
+	struct stat st;
+	if (force) {
+		unlink(filename);
+	} else {
+		if (stat(filename, &st) == 0) {
+			fprintf(stderr, "%s: %s: file exists\n", argv[0], filename);
+			exit(EXIT_FAILURE);
+		}
+	}
+	outfile->ostream.open(filename,std::ios::out | std::ios::binary);
+
+	char tmpname[strlen(tmpdir) + strlen("/pmtiles.XXXXXX") + 1];
+	sprintf(tmpname, "%s%s", tmpdir, "/pmtiles.XXXXXX");
+	int tmpfd = mkstemp(tmpname);
+	close(tmpfd);
+
+	outfile->tmptilesname = tmpname;
+	outfile->tilestmp.open(tmpname,std::ios::out | std::ios::binary);
+	outfile->offset = 0;
+	return outfile;
+}
+
+void serialize_string(const std::string &str, std::stringstream &ss) {
+	size_t MAX_SIZE = 10;
+	uint8_t slen;
+	slen = str.size();
+	if (slen > MAX_SIZE) {
+		fprintf(stderr, "%s: header value exceeds limit\n", str.data());
+		exit(EXIT_FAILURE);
+	}
+	ss.write((char *)&slen,1);
+	ss.write(str.data(),slen);
+
+	char zero = 0;
+	for (size_t i = 0; i < MAX_SIZE-slen; i++) {
+    ss.write(&zero,sizeof(char));
+	}
+}
+
+// TODO: endianness check
+std::string pmtilesv3_header::serialize() {
+	std::stringstream ss;
+	uint16_t MAGIC = 0x4d50;
+	ss.write((char *)&MAGIC,2);
+	uint16_t version = 2;
+	ss.write((char *)&version,2);
+
+	ss.write((char *)&root_dir_bytes,4);
+	ss.write((char *)&json_metadata_bytes,4);
+	ss.write((char *)&leaf_dirs_bytes,8);
+	ss.write((char *)&leaf_dirs_offset,8);
+	ss.write((char *)&tile_data_offset,8);
+	ss.write((char *)&addressed_tiles_count,8);
+	ss.write((char *)&tile_entries_count,8);
+	ss.write((char *)&unique_tile_contents_count,8);
+
+	uint8_t clustered_val = 0x0;
+	if (clustered) {
+		clustered_val = 0x1;
+	}
+	ss.write((char *)&clustered_val,1);
+
+	serialize_string(directory_compression,ss);
+	serialize_string(tile_compression,ss);
+	serialize_string(tile_format,ss);
+
+	ss.write((char *)&min_zoom,1);
+	ss.write((char *)&max_zoom,1);
+	ss.write((char *)&min_lon,4);
```

Comment:
Tippecanoe internally uses a `double` precision float for these, but right now the binary header design for pmtiles specifies a single-precision IEEE754 `float` for all floating-point header values - latitude, longitude. any footguns here?

Reply:
I replaced this with a fixed-precision 32 bit signed integer, the degree times 10e7 (10,000,000) which should resolve any roundtrip/arithmetic issues.